### PR TITLE
add memcached plugin

### PR DESCRIPTION
**Pull request by vetinari**
just allow stats() and get(KEY)

### CODE REVIEW

**Comment by vetinari** [comment]
:+1:

File: zmon_worker_monitor/builtins/plugins/memcached.py [hunk 1]
```diff
@@ -0,0 +1,131 @@
...
+            self.counter_factory = plugin_manager.get_plugin_obj_by_name('counter', 'Function')
+        return propartial(MemcachedWrapper, counter=self.counter_factory.create(factory_ctx), host=factory_ctx['host'])
+
+
+class MemcachedWrapper(object):
+    '''Class to allow only readonly access to underlying redis connection'''
+
+    def __init__(self, counter, host, port=11211, socket_connect_timeout=1):
+        if not host:
+            raise ConfigurationError('Memcached wrapper improperly configured. Valid redis host is required!')
```
Comment (mohabusama) [inline zmon_worker_monitor/builtins/plugins/memcached.py]: s/redis/memcached

File: zmon_worker_monitor/builtins/plugins/memcached.py [hunk 2]
```diff
@@ -0,0 +1,131 @@
...
+    def stats(self):
+        data = self.__con.stats()
+        ret = {}
+        for key in data:
+            if key in COUNTER_KEYS:
+                ret['{}_per_sec'.format(key.replace('total_', ''))] = \
+                    round(self._counter.key(key).per_second(data.get(key, 0)), 2)
+            else:
+                ret[key] = data[key]
+
```
Comment (mohabusama) [inline zmon_worker_monitor/builtins/plugins/memcached.py]: No return? :D

File: zmon_worker_monitor/builtins/plugins/memcached.py [hunk 3]
```diff
@@ -0,0 +1,131 @@
...
+# -*- coding: utf-8 -*-
+
+from pymemcache.client.base import Client
+
+from zmon_worker_monitor.zmon_worker.errors import ConfigurationError
+
+from zmon_worker_monitor.adapters.ifunctionfactory_plugin import IFunctionFactoryPlugin, propartial
+from zmon_worker_monitor import plugin_manager
+
+COUNTER_KEYS = frozenset([
```
Comment (mohabusama) [inline zmon_worker_monitor/builtins/plugins/memcached.py]: Do we need all those keys by default?

File: zmon_worker_monitor/builtins/plugins/memcached.worker_plugin [hunk 4]
```diff
@@ -0,0 +1,17 @@
+[Core]
+Name = memcached
+Module = memcached
+
+[Documentation]
+Author = John Doe
+Version = 0.1
+Website = https://github.com/zalando/zmon-worker
```
Comment (mohabusama) [inline zmon_worker_monitor/builtins/plugins/memcached.worker_plugin]: s/zalando/zalando-zmon
Author also is useful ;)

**Comment by mohabusama** [comment]
Please add tests

File: zmon_worker_monitor/builtins/plugins/memcached.py (hunk 3 above)
Comment (vetinari) [reply zmon_worker_monitor/builtins/plugins/memcached.py]: probably not ... checking which ones should be in by default

File: zmon_worker_monitor/builtins/plugins/memcached.py (hunk 2 above)
Comment (vetinari) [reply zmon_worker_monitor/builtins/plugins/memcached.py]: oops :)

File: tests/test_memcached.py [hunk 5]
```diff
@@ -0,0 +1,11 @@
+import pytest
+
+from zmon_worker_monitor.builtins.plugins.memcached import MemcachedWrapper, ConfigurationError
+
+
+def test_memcached_config_error():
```
Comment (mohabusama) [inline tests/test_memcached.py]: The logic in ``stats()`` is worth adding another test imho 😉

**Comment by vetinari** [comment]
:+1:

File: zmon_worker_monitor/builtins/plugins/memcached.py [hunk 6]
```diff
@@ -0,0 +1,115 @@
...
+        :return: an object that implements a check function
+        """
+        # load plugins dependencies and store them locally for efficiency
+        if not self.counter_factory:
+            self.counter_factory = plugin_manager.get_plugin_obj_by_name('counter', 'Function')
+        return propartial(MemcachedWrapper, counter=self.counter_factory.create(factory_ctx), host=factory_ctx['host'])
+
+
+class MemcachedWrapper(object):
+    '''Class to allow only readonly access to underlying redis connection'''
```
Comment (mohabusama) [inline zmon_worker_monitor/builtins/plugins/memcached.py]: s/only readonly/readonly
s/redis/memcached

**Comment by vetinari** [comment]
:+1:

**Comment by mohabusama** [comment]
👍

**Pull request merged**